File: services/gradebook-svc/src/services/gradebook.service.ts
/**
 * Gradebook Service
 *
 * Handles all gradebook operations including:
 * - Assignment management
 * - Grade calculations
 * - Grade categories
 * - Grade history and audit trail
 * - Bulk import/export
 */

import { PrismaClient } from '../generated/prisma-client/index.js';
import type {
  Assignment,
  Grade,
  GradeCategory,
  GradebookConfig,
  GradeAuditLog,
  AssignmentType,
  GradeStatus,
  GradeCalculationType
} from '../generated/prisma-client/index.js';

const prisma = new PrismaClient();

export interface CreateAssignmentInput {
  gradebookConfigId: string;
  categoryId?: string;
  title: string;
  description?: string;
  type: AssignmentType;
  totalPoints: number;
  extraCredit?: boolean;
  dueDate?: Date;
  assignedDate?: Date;
  availableFrom?: Date;
  availableUntil?: Date;
  allowLateSubmissions?: boolean;
  latePenaltyPercent?: number;
  assessmentId?: string;
  createdBy: string;
}

export interface UpdateGradeInput {
  gradeId: string;
  score?: number | null;
  status?: GradeStatus;
  feedback?: string;
  privateNotes?: string;
  rubricScores?: any;
  gradedBy: string;
  overrideReason?: string;
}

export interface BulkGradeImport {
  assignmentId: string;
  grades: Array<{
    studentId: string;
    score: number | null;
    feedback?: string;
  }>;
  importedBy: string;
}

export interface GradebookCalculation {
  studentId: string;
  categoryGrades: Array<{
    categoryId: string;
    categoryName: string;
    score: number;
    weight: number;
  }>;
  overallGrade: number;
  letterGrade: string;
  totalPoints: number;
  earnedPoints: number;
}

export class GradebookService {
  /**
   * Get gradebook configuration for a classroom
   */
  async getGradebookConfig(classroomId: string): Promise<GradebookConfig | null> {
    return prisma.gradebookConfig.findUnique({
      where: { classroomId },
      include: {
        categories: {
          orderBy: { orderIndex: 'asc' }
        },
        assignments: {
          where: { status: 'PUBLISHED' },
          orderBy: { dueDate: 'desc' },
          include: {
            category: true,
            grades: true
          }
        }
      }
    });
  }

  /**
   * Create or update gradebook configuration
   */
  async upsertGradebookConfig(
    classroomId: string,
    teacherId: string,
    tenantId: string,
    config: Partial<GradebookConfig>
  ): Promise<GradebookConfig> {
    return prisma.gradebookConfig.upsert({
      where: { classroomId },
      create: {
        classroomId,
        teacherId,
        tenantId,
        ...config
      },
      update: config,
      include: {
        categories: true
      }
    });
  }

  /**
   * Create grade category
   */
  async createCategory(
    gradebookConfigId: string,
    name: string,
    weight: number,
    options?: {
      color?: string;
      dropLowest?: number;
      orderIndex?: number;
    }
  ): Promise<GradeCategory> {
    return prisma.gradeCategory.create({
      data: {
        gradebookConfigId,
        name,
        weight,
        color: options?.color,
        dropLowest: options?.dropLowest ?? 0,
        orderIndex: options?.orderIndex ?? 0
      }
    });
  }

  /**
   * Update category weights
   */
  async updateCategoryWeights(
    categories: Array<{ id: string; weight: number }>
  ): Promise<void> {
    await prisma.$transaction(
      categories.map(cat =>
        prisma.gradeCategory.update({
          where: { id: cat.id },
          data: { weight: cat.weight }
        })
      )
    );
  }

  /**
   * Create assignment
   */
  async createAssignment(input: CreateAssignmentInput): Promise<Assignment> {
    const config = await prisma.gradebookConfig.findUnique({
      where: { id: input.gradebookConfigId }
    });

    if (!config) {
      throw new Error('Gradebook config not found');
    }

    return prisma.assignment.create({
      data: {
        tenantId: config.tenantId,
        gradebookConfigId: input.gradebookConfigId,
        categoryId: input.categoryId,
        title: input.title,
        description: input.description,
        type: input.type,
        totalPoints: input.totalPoints,
        extraCredit: input.extraCredit ?? false,
        dueDate: input.dueDate,
        assignedDate: input.assignedDate,
        availableFrom: input.availableFrom,
        availableUntil: input.availableUntil,
        allowLateSubmissions: input.allowLateSubmissions ?? true,
        latePenaltyPercent: input.latePenaltyPercent,
        assessmentId: input.assessmentId,
        createdBy: input.createdBy,
        status: 'DRAFT'
      },
      include: {
        category: true
      }
    });
  }

  /**
   * Update assignment
   */
  async updateAssignment(
    assignmentId: string,
    updates: Partial<Assignment>
  ): Promise<Assignment> {
    return prisma.assignment.update({
      where: { id: assignmentId },
      data: updates,
      include: {
        category: true,
        grades: true
      }
    });
  }

  /**
   * Publish assignment
   */
  async publishAssignment(assignmentId: string): Promise<Assignment> {
    return prisma.assignment.update({
      where: { id: assignmentId },
      data: {
        status: 'PUBLISHED',
        publishedAt: new Date()
      }
    });
  }

  /**
   * Delete assignment
   */
  async deleteAssignment(assignmentId: string): Promise<void> {
    await prisma.assignment.delete({
      where: { id: assignmentId }
    });
  }

  /**
   * Submit or update a grade
   */
  async submitGrade(input: UpdateGradeInput): Promise<Grade> {
    const { gradeId, score, status, feedback, privateNotes, rubricScores, gradedBy, overrideReason } = input;

    // Get existing grade
    const existingGrade = await prisma.grade.findUnique({
      where: { id: gradeId },
      include: { assignment: true }
    });

    if (!existingGrade) {
      throw new Error('Grade not found');
    }

    // Calculate percentage
    const maxPoints = existingGrade.assignment.totalPoints;
    const percentage = score !== null && score !== undefined ? (score / maxPoints) * 100 : null;

    // Determine if this is an override
    const isOverride = overrideReason !== undefined && overrideReason !== null;
    const originalScore = isOverride ? existingGrade.score : existingGrade.originalScore;

    // Update grade
    const updatedGrade = await prisma.grade.update({
      where: { id: gradeId },
      data: {
        score,
        maxPoints,
        percentage,
        status: status ?? (score !== null ? 'GRADED' : existingGrade.status),
        feedback,
        privateNotes,
        rubricScores,
        gradedBy,
        gradedAt: new Date(),
        isOverride,
        overrideReason,
        originalScore
      }
    });

    // Create audit log entry
    await this.createAuditLog({
      gradeId,
      assignmentId: existingGrade.assignmentId,
      studentId: existingGrade.studentId,
      action: isOverride ? 'override_applied' : 'updated',
      fieldChanged: 'score',
      oldValue: existingGrade.score?.toString(),
      newValue: score?.toString(),
      reason: overrideReason,
      changedBy: gradedBy
    });

    return updatedGrade;
  }

  /**
   * Create grade for student (if not exists)
   */
  async getOrCreateGrade(
    assignmentId: string,
    studentId: string,
    tenantId: string
  ): Promise<Grade> {
    const assignment = await prisma.assignment.findUnique({
      where: { id: assignmentId }
    });

    if (!assignment) {
      throw new Error('Assignment not found');
    }

    return prisma.grade.upsert({
      where: {
        assignmentId_studentId: {
          assignmentId,
          studentId
        }
      },
      create: {
        assignmentId,
        studentId,
        tenantId,
        maxPoints: assignment.totalPoints,
        status: 'NOT_SUBMITTED'
      },
      update: {}
    });
  }

  /**
   * Bulk import grades
   */
  async bulkImportGrades(input: BulkGradeImport): Promise<{
    success: number;
    failed: number;
    errors: Array<{ studentId: string; error: string }>
  }> {
    const { assignmentId, grades, importedBy } = input;
    let success = 0;
    let failed = 0;
    const errors: Array<{ studentId: string; error: string }> = [];

    const assignment = await prisma.assignment.findUnique({
      where: { id: assignmentId }
    });

    if (!assignment) {
      throw new Error('Assignment not found');
    }

    for (const gradeData of grades) {
      try {
        const grade = await this.getOrCreateGrade(
          assignmentId,
          gradeData.studentId,
          assignment.tenantId
        );

        await this.submitGrade({
          gradeId: grade.id,
          score: gradeData.score,
          feedback: gradeData.feedback,
          gradedBy: importedBy
        });

        success++;
      } catch (error) {
        failed++;
        errors.push({
          studentId: gradeData.studentId,
          error: error instanceof Error ? error.message : 'Unknown error'
        });
      }
    }

    return { success, failed, errors };
  }

  /**
   * Calculate grades for a student
   */
  async calculateStudentGrades(
    gradebookConfigId: string,
    studentId: string
  ): Promise<GradebookCalculation> {
    const config = await prisma.gradebookConfig.findUnique({
      where: { id: gradebookConfigId },
      include: {
        categories: true,
        assignments: {
          where: {
            status: 'PUBLISHED',
            extraCredit: false
          },
          include: {
            category: true,
            grades: {
              where: { studentId }
            }
          }
        }
      }
    });

    if (!config) {
      throw new Error('Gradebook config not found');
    }

    const categoryGrades: Array<{
      categoryId: string;
      categoryName: string;
      score: number;
      weight: number;
    }> = [];

    let totalPoints = 0;
    let earnedPoints = 0;

    if (config.calculationType === 'WEIGHTED_CATEGORIES') {
      // Calculate weighted category grades
      for (const category of config.categories) {
        const categoryAssignments = config.assignments.filter(
          a => a.categoryId === category.id
        );

        if (categoryAssignments.length === 0) continue;

        let categoryTotal = 0;
        let categoryEarned = 0;
        let validGrades = 0;

        for (const assignment of categoryAssignments) {
          const grade = assignment.grades[0];
          if (grade?.score !== null && grade?.score !== undefined) {
            categoryTotal += assignment.totalPoints;
            categoryEarned += grade.score;
            validGrades++;
          }
        }

        if (validGrades > 0) {
          const categoryScore = (categoryEarned / categoryTotal) * 100;
          categoryGrades.push({
            categoryId: category.id,
            categoryName: category.name,
            score: categoryScore,
            weight: category.weight
          });
        }
      }

      // Calculate weighted overall grade
      const totalWeight = categoryGrades.reduce((sum, cat) => sum + cat.weight, 0);
      const weightedScore = categoryGrades.reduce(
        (sum, cat) => sum + (cat.score * cat.weight),
        0
      );
      const overallGrade = totalWeight > 0 ? weightedScore / totalWeight : 0;

      return {
        studentId,
        categoryGrades,
        overallGrade,
        letterGrade: this.getLetterGrade(overallGrade, config.letterGradeScale as any),
        totalPoints,
        earnedPoints
      };
    } else {
      // TOTAL_POINTS calculation
      for (const assignment of config.assignments) {
        const grade = assignment.grades[0];
        totalPoints += assignment.totalPoints;
        if (grade?.score !== null && grade?.score !== undefined) {
          earnedPoints += grade.score;
        }
      }

      const overallGrade = totalPoints > 0 ? (earnedPoints / totalPoints) * 100 : 0;

      return {
        studentId,
        categoryGrades: [],
        overallGrade,
        letterGrade: this.getLetterGrade(overallGrade, config.letterGradeScale as any),
        totalPoints,
        earnedPoints
      };
    }
  }

  /**
   * Get student grades
   */
  async getStudentGrades(
    classroomId: string,
    studentId: string
  ): Promise<{
    grades: Grade[];
    calculation: GradebookCalculation;
  }> {
    const config = await this.getGradebookConfig(classroomId);
    if (!config) {
      throw new Error('Gradebook config not found');
    }

    const grades = await prisma.grade.findMany({
      where: {
        studentId,
        assignment: {
          gradebookConfigId: config.id,
          status: 'PUBLISHED'
        }
      },
      include: {
        assignment: {
          include: {
            category: true
          }
        }
      },
      orderBy: {
        assignment: {
          dueDate: 'desc'
        }
      }
    });

    const calculation = await this.calculateStudentGrades(config.id, studentId);

    return { grades, calculation };
  }

  /**
   * Get classroom gradebook
   */
  async getClassroomGradebook(classroomId: string): Promise<{
    config: GradebookConfig;
    students: Array<{
      studentId: string;
      calculation: GradebookCalculation;
    }>;
  }> {
    const config = await this.getGradebookConfig(classroomId);
    if (!config) {
      throw new Error('Gradebook config not found');
    }

    // Get all student IDs from grades
    const studentIds = await prisma.grade.findMany({
      where: {
        assignment: {
          gradebookConfigId: config.id
        }
      },
      select: {
        studentId: true
      },
      distinct: ['studentId']
    });

    const students = await Promise.all(
      studentIds.map(async ({ studentId }) => ({
        studentId,
        calculation: await this.calculateStudentGrades(config.id, studentId)
      }))
    );

    return { config, students };
  }

  /**
   * Export gradebook to CSV
   */
  async exportGradebook(classroomId: string): Promise<string> {
    const { config, students } = await this.getClassroomGradebook(classroomId);

    // Generate CSV headers
    const headers = ['Student ID', 'Overall Grade', 'Letter Grade'];
    config.assignments.forEach(assignment => {
      headers.push(`${assignment.title} (${assignment.totalPoints}pts)`);
    });

    // Generate CSV rows
    const rows = students.map(student => {
      const row = [
        student.studentId,
        student.calculation.overallGrade.toFixed(2),
        student.calculation.letterGrade
      ];

      config.assignments.forEach(assignment => {
        const grade = assignment.grades.find(g => g.studentId === student.studentId);
        row.push(grade?.score?.toString() ?? '');
      });

      return row.join(',');
    });

    return [headers.join(','), ...rows].join('\n');
  }

  /**
   * Get grade history (audit log)
   */
  async getGradeHistory(gradeId: string): Promise<GradeAuditLog[]> {
    return prisma.gradeAuditLog.findMany({
      where: { gradeId },
      orderBy: { changedAt: 'desc' }
    });
  }

  /**
   * Create audit log entry
   */
  private async createAuditLog(data: {
    gradeId: string;
    assignmentId: string;
    studentId: string;
    action: string;
    fieldChanged?: string;
    oldValue?: string;
    newValue?: string;
    reason?: string;
    changedBy: string;
  }): Promise<GradeAuditLog> {
    return prisma.gradeAuditLog.create({
      data: {
        gradeId: data.gradeId,
        assignmentId: data.assignmentId,
        studentId: data.studentId,
        action: data.action,
        fieldChanged: data.fieldChanged,
        oldValue: data.oldValue,
        newValue: data.newValue,
        reason: data.reason,
        changedBy: data.changedBy
      }
    });
  }

  /**
   * Get letter grade from percentage
   */
  private getLetterGrade(
    percentage: number,
    scale: Record<string, number>
  ): string {
    const grades = Object.entries(scale).sort((a, b) => b[1] - a[1]);
    for (const [grade, threshold] of grades) {
      if (percentage >= threshold) {
        return grade;
      }
    }
    return 'F';
  }
}

export const gradebookService = new GradebookService();

File: libs/flutter-common/lib/services/gradebook_service.dart
/// Gradebook Service
///
/// API client for gradebook operations in mobile-teacher app.
library;

import 'package:flutter_common/api/api_client.dart';

/// Service for gradebook API operations.
class GradebookService {
  final AivoApiClient _client;

  GradebookService({AivoApiClient? client})
      : _client = client ?? AivoApiClient.instance;

  static const String _basePath = '/gradebook-svc/gradebook';

  // ═══════════════════════════════════════════════════════════════════════════
  // GRADEBOOK DATA
  // ═══════════════════════════════════════════════════════════════════════════

  /// Fetch the complete gradebook for a classroom.
  Future<GradebookData> getClassroomGradebook(String classroomId) async {
    final response = await _client.get<Map<String, dynamic>>(
      '$_basePath/classroom/$classroomId',
    );
    return GradebookData.fromJson(response.data!);
  }

  /// Get gradebook configuration for a classroom.
  Future<GradebookConfig?> getGradebookConfig(String classroomId) async {
    try {
      final response = await _client.get<Map<String, dynamic>>(
        '$_basePath/config/$classroomId',
      );
      return GradebookConfig.fromJson(response.data!);
    } catch (e) {
      return null;
    }
  }

  /// Create or update gradebook configuration.
  Future<GradebookConfig> upsertGradebookConfig({
    required String classroomId,
    required String teacherId,
    required String tenantId,
    String? gradingScale,
    bool? showOverallGrade,
    bool? allowLateSubmissions,
    int? latePenaltyPercent,
  }) async {
    final response = await _client.post<Map<String, dynamic>>(
      '$_basePath/config',
      data: {
        'classroomId': classroomId,
        'teacherId': teacherId,
        'tenantId': tenantId,
        if (gradingScale != null) 'gradingScale': gradingScale,
        if (showOverallGrade != null) 'showOverallGrade': showOverallGrade,
        if (allowLateSubmissions != null)
          'allowLateSubmissions': allowLateSubmissions,
        if (latePenaltyPercent != null)
          'latePenaltyPercent': latePenaltyPercent,
      },
    );
    return GradebookConfig.fromJson(response.data!);
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // CATEGORIES
  // ═══════════════════════════════════════════════════════════════════════════

  /// Create a grade category.
  Future<GradeCategory> createCategory({
    required String gradebookConfigId,
    required String name,
    required double weight,
    String? color,
    int? dropLowest,
    int? orderIndex,
  }) async {
    final response = await _client.post<Map<String, dynamic>>(
      '$_basePath/categories',
      data: {
        'gradebookConfigId': gradebookConfigId,
        'name': name,
        'weight': weight,
        if (color != null) 'color': color,
        if (dropLowest != null) 'dropLowest': dropLowest,
        if (orderIndex != null) 'orderIndex': orderIndex,
      },
    );
    return GradeCategory.fromJson(response.data!);
  }

  /// Update category weights.
  Future<void> updateCategoryWeights(
      List<Map<String, dynamic>> categories) async {
    await _client.put<Map<String, dynamic>>(
      '$_basePath/categories/weights',
      data: {'categories': categories},
    );
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // ASSIGNMENTS
  // ═══════════════════════════════════════════════════════════════════════════

  /// Create an assignment.
  Future<GradebookAssignment> createAssignment({
    required String classroomId,
    required String categoryId,
    required String teacherId,
    required String tenantId,
    required String title,
    required double maxPoints,
    DateTime? dueDate,
    String? description,
  }) async {
    final response = await _client.post<Map<String, dynamic>>(
      '$_basePath/assignments',
      data: {
        'classroomId': classroomId,
        'categoryId': categoryId,
        'teacherId': teacherId,
        'tenantId': tenantId,
        'title': title,
        'maxPoints': maxPoints,
        if (dueDate != null) 'dueDate': dueDate.toIso8601String(),
        if (description != null) 'description': description,
      },
    );
    return GradebookAssignment.fromJson(response.data!);
  }

  /// Update an assignment.
  Future<GradebookAssignment> updateAssignment(
    String assignmentId,
    Map<String, dynamic> updates,
  ) async {
    final response = await _client.put<Map<String, dynamic>>(
      '$_basePath/assignments/$assignmentId',
      data: updates,
    );
    return GradebookAssignment.fromJson(response.data!);
  }

  /// Publish an assignment.
  Future<GradebookAssignment> publishAssignment(String assignmentId) async {
    final response = await _client.post<Map<String, dynamic>>(
      '$_basePath/assignments/$assignmentId/publish',
    );
    return GradebookAssignment.fromJson(response.data!);
  }

  /// Delete an assignment.
  Future<void> deleteAssignment(String assignmentId) async {
    await _client.delete<void>('$_basePath/assignments/$assignmentId');
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // GRADES
  // ═══════════════════════════════════════════════════════════════════════════

  /// Submit a grade.
  Future<StudentGrade> submitGrade({
    required String assignmentId,
    required String studentId,
    required String classroomId,
    required String gradedBy,
    required String tenantId,
    double? pointsEarned,
    String? letterGrade,
    String? feedback,
    bool? isExcused,
    bool? isLate,
    bool? isMissing,
  }) async {
    final response = await _client.post<Map<String, dynamic>>(
      '$_basePath/grades',
      data: {
        'assignmentId': assignmentId,
        'studentId': studentId,
        'classroomId': classroomId,
        'gradedBy': gradedBy,
        'tenantId': tenantId,
        if (pointsEarned != null) 'pointsEarned': pointsEarned,
        if (letterGrade != null) 'letterGrade': letterGrade,
        if (feedback != null) 'feedback': feedback,
        if (isExcused != null) 'isExcused': isExcused,
        if (isLate != null) 'isLate': isLate,
        if (isMissing != null) 'isMissing': isMissing,
      },
    );
    return StudentGrade.fromJson(response.data!);
  }

  /// Update a grade.
  Future<StudentGrade> updateGrade(
    String gradeId,
    Map<String, dynamic> updates,
  ) async {
    final response = await _client.put<Map<String, dynamic>>(
      '$_basePath/grades/$gradeId',
      data: updates,
    );
    return StudentGrade.fromJson(response.data!);
  }

  /// Get grade history for audit trail.
  Future<List<GradeHistoryEntry>> getGradeHistory(String gradeId) async {
    final response = await _client.get<List<dynamic>>(
      '$_basePath/grades/$gradeId/history',
    );
    return (response.data!)
        .map((e) => GradeHistoryEntry.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  /// Get all grades for a student in a classroom.
  Future<StudentGradesResult> getStudentGrades(
    String classroomId,
    String studentId,
  ) async {
    final response = await _client.get<Map<String, dynamic>>(
      '$_basePath/student/$studentId/classroom/$classroomId',
    );
    return StudentGradesResult.fromJson(response.data!);
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // IMPORT/EXPORT
  // ═══════════════════════════════════════════════════════════════════════════

  /// Bulk import grades.
  Future<BulkImportResult> bulkImportGrades({
    required String classroomId,
    required String teacherId,
    required String tenantId,
    required List<Map<String, dynamic>> grades,
  }) async {
    final response = await _client.post<Map<String, dynamic>>(
      '$_basePath/import',
      data: {
        'classroomId': classroomId,
        'teacherId': teacherId,
        'tenantId': tenantId,
        'grades': grades,
      },
    );
    return BulkImportResult.fromJson(response.data!);
  }

  /// Export gradebook as CSV.
  Future<String> exportGradebook(String classroomId) async {
    final response = await _client.get<String>(
      '$_basePath/export/$classroomId',
    );
    return response.data!;
  }
}

// ═══════════════════════════════════════════════════════════════════════════
// DATA MODELS
// ═══════════════════════════════════════════════════════════════════════════

/// Complete gradebook data for a classroom.
class GradebookData {
  final String classroomId;
  final String classroomName;
  final GradebookConfig? config;
  final List<GradebookStudent> students;
  final List<GradebookAssignment> assignments;
  final List<GradeCategory> categories;

  GradebookData({
    required this.classroomId,
    required this.classroomName,
    this.config,
    required this.students,
    required this.assignments,
    required this.categories,
  });

  factory GradebookData.fromJson(Map<String, dynamic> json) {
    return GradebookData(
      classroomId: json['classroomId'] as String,
      classroomName: json['classroomName'] as String? ?? '',
      config: json['config'] != null
          ? GradebookConfig.fromJson(json['config'] as Map<String, dynamic>)
          : null,
      students: (json['students'] as List<dynamic>?)
              ?.map((e) =>
                  GradebookStudent.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      assignments: (json['assignments'] as List<dynamic>?)
              ?.map((e) =>
                  GradebookAssignment.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      categories: (json['categories'] as List<dynamic>?)
              ?.map((e) => GradeCategory.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

/// Gradebook configuration.
class GradebookConfig {
  final String id;
  final String classroomId;
  final String gradingScale;
  final bool showOverallGrade;
  final bool allowLateSubmissions;
  final int latePenaltyPercent;

  GradebookConfig({
    required this.id,
    required this.classroomId,
    required this.gradingScale,
    required this.showOverallGrade,
    required this.allowLateSubmissions,
    required this.latePenaltyPercent,
  });

  factory GradebookConfig.fromJson(Map<String, dynamic> json) {
    return GradebookConfig(
      id: json['id'] as String,
      classroomId: json['classroomId'] as String,
      gradingScale: json['gradingScale'] as String? ?? 'POINTS',
      showOverallGrade: json['showOverallGrade'] as bool? ?? true,
      allowLateSubmissions: json['allowLateSubmissions'] as bool? ?? true,
      latePenaltyPercent: json['latePenaltyPercent'] as int? ?? 0,
    );
  }
}

/// Grade category.
class GradeCategory {
  final String id;
  final String name;
  final double weight;
  final String? color;
  final int dropLowest;
  final int orderIndex;

  GradeCategory({
    required this.id,
    required this.name,
    required this.weight,
    this.color,
    required this.dropLowest,
    required this.orderIndex,
  });

  factory GradeCategory.fromJson(Map<String, dynamic> json) {
    return GradeCategory(
      id: json['id'] as String,
      name: json['name'] as String,
      weight: (json['weight'] as num).toDouble(),
      color: json['color'] as String?,
      dropLowest: json['dropLowest'] as int? ?? 0,
      orderIndex: json['orderIndex'] as int? ?? 0,
    );
  }
}

/// Student in gradebook.
class GradebookStudent {
  final String id;
  final String displayName;
  final String? avatarUrl;
  final double? overallGrade;
  final String? letterGrade;
  final Map<String, StudentGrade> grades;

  GradebookStudent({
    required this.id,
    required this.displayName,
    this.avatarUrl,
    this.overallGrade,
    this.letterGrade,
    required this.grades,
  });

  factory GradebookStudent.fromJson(Map<String, dynamic> json) {
    final gradesJson = json['grades'] as Map<String, dynamic>? ?? {};
    return GradebookStudent(
      id: json['id'] as String,
      displayName: json['displayName'] as String,
      avatarUrl: json['avatarUrl'] as String?,
      overallGrade: (json['overallGrade'] as num?)?.toDouble(),
      letterGrade: json['letterGrade'] as String?,
      grades: gradesJson.map(
        (k, v) => MapEntry(k, StudentGrade.fromJson(v as Map<String, dynamic>)),
      ),
    );
  }
}

/// Assignment in gradebook.
class GradebookAssignment {
  final String id;
  final String title;
  final String? description;
  final String categoryId;
  final double maxPoints;
  final DateTime? dueDate;
  final bool isPublished;
  final DateTime createdAt;

  GradebookAssignment({
    required this.id,
    required this.title,
    this.description,
    required this.categoryId,
    required this.maxPoints,
    this.dueDate,
    required this.isPublished,
    required this.createdAt,
  });

  factory GradebookAssignment.fromJson(Map<String, dynamic> json) {
    return GradebookAssignment(
      id: json['id'] as String,
      title: json['title'] as String,
      description: json['description'] as String?,
      categoryId: json['categoryId'] as String,
      maxPoints: (json['maxPoints'] as num).toDouble(),
      dueDate: json['dueDate'] != null
          ? DateTime.parse(json['dueDate'] as String)
          : null,
      isPublished: json['isPublished'] as bool? ?? false,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }
}

/// Individual grade entry.
class StudentGrade {
  final String id;
  final String assignmentId;
  final String studentId;
  final double? pointsEarned;
  final String? letterGrade;
  final double? percentage;
  final String? feedback;
  final bool isExcused;
  final bool isLate;
  final bool isMissing;
  final DateTime? gradedAt;

  StudentGrade({
    required this.id,
    required this.assignmentId,
    required this.studentId,
    this.pointsEarned,
    this.letterGrade,
    this.percentage,
    this.feedback,
    required this.isExcused,
    required this.isLate,
    required this.isMissing,
    this.gradedAt,
  });

  factory StudentGrade.fromJson(Map<String, dynamic> json) {
    return StudentGrade(
      id: json['id'] as String,
      assignmentId: json['assignmentId'] as String,
      studentId: json['studentId'] as String,
      pointsEarned: (json['pointsEarned'] as num?)?.toDouble(),
      letterGrade: json['letterGrade'] as String?,
      percentage: (json['percentage'] as num?)?.toDouble(),
      feedback: json['feedback'] as String?,
      isExcused: json['isExcused'] as bool? ?? false,
      isLate: json['isLate'] as bool? ?? false,
      isMissing: json['isMissing'] as bool? ?? false,
      gradedAt: json['gradedAt'] != null
          ? DateTime.parse(json['gradedAt'] as String)
          : null,
    );
  }
}

/// Grade history entry for audit trail.
class GradeHistoryEntry {
  final String id;
  final String gradeId;
  final double? previousPoints;
  final double? newPoints;
  final String? previousLetter;
  final String? newLetter;
  final String changedBy;
  final String changeReason;
  final DateTime changedAt;

  GradeHistoryEntry({
    required this.id,
    required this.gradeId,
    this.previousPoints,
    this.newPoints,
    this.previousLetter,
    this.newLetter,
    required this.changedBy,
    required this.changeReason,
    required this.changedAt,
  });

  factory GradeHistoryEntry.fromJson(Map<String, dynamic> json) {
    return GradeHistoryEntry(
      id: json['id'] as String,
      gradeId: json['gradeId'] as String,
      previousPoints: (json['previousPoints'] as num?)?.toDouble(),
      newPoints: (json['newPoints'] as num?)?.toDouble(),
      previousLetter: json['previousLetter'] as String?,
      newLetter: json['newLetter'] as String?,
      changedBy: json['changedBy'] as String,
      changeReason: json['changeReason'] as String? ?? '',
      changedAt: DateTime.parse(json['changedAt'] as String),
    );
  }
}

/// Result for student grades query.
class StudentGradesResult {
  final String studentId;
  final String classroomId;
  final double? overallGrade;
  final String? letterGrade;
  final List<StudentGrade> grades;
  final Map<String, double> categoryGrades;

  StudentGradesResult({
    required this.studentId,
    required this.classroomId,
    this.overallGrade,
    this.letterGrade,
    required this.grades,
    required this.categoryGrades,
  });

  factory StudentGradesResult.fromJson(Map<String, dynamic> json) {
    final catGrades = json['categoryGrades'] as Map<String, dynamic>? ?? {};
    return StudentGradesResult(
      studentId: json['studentId'] as String,
      classroomId: json['classroomId'] as String,
      overallGrade: (json['overallGrade'] as num?)?.toDouble(),
      letterGrade: json['letterGrade'] as String?,
      grades: (json['grades'] as List<dynamic>?)
              ?.map((e) => StudentGrade.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      categoryGrades:
          catGrades.map((k, v) => MapEntry(k, (v as num).toDouble())),
    );
  }
}

/// Result of bulk import operation.
class BulkImportResult {
  final int imported;
  final int updated;
  final int failed;
  final List<String> errors;

  BulkImportResult({
    required this.imported,
    required this.updated,
    required this.failed,
    required this.errors,
  });

  factory BulkImportResult.fromJson(Map<String, dynamic> json) {
    return BulkImportResult(
      imported: json['imported'] as int? ?? 0,
      updated: json['updated'] as int? ?? 0,
      failed: json['failed'] as int? ?? 0,
      errors: (json['errors'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
    );
  }
}
